### ISSUE TITLE
Shorten use of ANY_DISCONTINUOUS_SPACE and ANY_SPACE in PSy layer and kernel stub generator

### ISSUE THREAD

**Issue by TeranIvy** [stfc/PSyclone#733]
Dynamo0.3 API uses full function-space metadata names to create unique variable names in PSy layer and kernel-stub generator. This may lead to problems and too long variable names. A specific example from the generated code that creates compilation error with both Gnu and Intel compilers is
```
algorithm/runtime_constants_mod_psy.f90(2537): error #6439: This symbol has too many characters.   [BASIS_ANY_SPACE_9_CHI_XYZ_ON_ANY_DISCONTINU]
      REAL(KIND=r_def), allocatable :: basis_any_space_9_chi_xyz_on_any_discontinuous_space_1_latitude_w3(:,:,:)
```
Ben Shipway believes that the maximum variable name in Fortran is 63 characters. However, the truncation above indicates 43. This needs to be checked.

To overcome this issue it was suggested to leave the function space metadata as they are but use short version of their names in PSy layer and kernel-stub generator. Current suggestion for use of `ANY_DISCONTINUOUS_SPACE` metadata in code generation is `any_d_space`, but better ideas are welcome. There will be a separate issue for a short name for `ANY_SPACE` metadata when generating code.

**Comment by TeranIvy** [comment]
Assigning @AdamVoysey as he volunteered to implement the change, @sergisiso for advising on Symbol names and @stevemullerworth to be able to see the progress.

**Comment by AdamVoysey** [comment]
The Fortran 2003 standard is quite clear on the maximum name length:

> C301 (R304) The maximum length of a name is 63 characters.

This is constraint C301 (R303) and C601 (R603) in Fortran 2008 and Fortran 2018 respectively, but the wording is otherwise unchanged. This limit was previously 31 in Fortran 95.

There is no obvious reason why this should be truncated at 43 characters in the error message - this is probably just one of `ifort`'s idiosyncrasies. The compiler documentation specifically says it implements the limit as 63.

**Comment by TeranIvy** [comment]
Thanks for checking, @AdamVoysey. I think there are 3 separate issues to enforce this properly:
1. This issue to use short name for `ANY_DISCONTINUOUS_SPACE` metadata when creating argument names,
2. Issue to use short name for `ANY_SPACE` metadata when creating argument names (not yet created),
3. Issue to limit argument names to 63 characters (not yet created).

I would argue that we need to complete 1. and 2. before limiting lengths of argument names. Regarding shortened names for `any_*_spaces`, I think using `aspace` and `adspace` for code generation would be descriptive enough (and not that people look into generated code much). It would also follow the logic of using function space names which do not have underscores in them such as `diff_basis_w1_on_w0`.

We would need this for the next PSyclone release scheduled for this May. Would you like to split the work? We can work in parallel, you on `ANY_DISCONTINOUOUS_SPACE` and I on `ANY_SPACE` and then we decide who takes character limit.

**Comment by TeranIvy** [comment]
Hi @AdamVoysey, I will take this on as we need it for the upcoming May release and the code review and merging process usually takes 2-3 weeks. It is not really fair to burden a new developer with something that is urgent. There are plenty of longer-term LFRic issues that are still open and simpler to do than this one. I am happy to point you to them.

**Comment by TeranIvy** [comment]
Added handling of `ANY_SPACE` it  is handled within the same constructs and functions as `ANY_DISCONTINUOUS_SPACE`  (`function_space.mangled_name` and `function_space.orig_name` in `dynamo0p3.py`).
